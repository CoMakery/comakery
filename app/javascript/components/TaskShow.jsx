import React from 'react'
import PropTypes from 'prop-types'
import Layout from './layouts/Layout'
import {fetch as fetchPolyfill} from 'whatwg-fetch'
import {Decimal} from 'decimal.js'
import InputFieldDropdown from './styleguide/InputFieldDropdown'
import InputFieldWhiteDark from './styleguide/InputFieldWhiteDark'
import InputFieldDescriptionMiddle from './styleguide/InputFieldDescriptionMiddle'
import Button from './styleguide/Button'
import ButtonBorder from './styleguide/ButtonBorder'
import Flash from './layouts/Flash'

class TaskShow extends React.Component {
  constructor(props) {
    super(props)

    this.goBack = this.goBack.bind(this)
    this.errorAdd = this.errorAdd.bind(this)
    this.errorRemove = this.errorRemove.bind(this)
    this.disable = this.disable.bind(this)
    this.enable = this.enable.bind(this)
    this.handleFieldChange = this.handleFieldChange.bind(this)
    this.fetchRecipientAddress = this.fetchRecipientAddress.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)

    this.state = {
      flashMessages     : [],
      errors            : {},
      disabled          : {},
      formAction        : this.props.formAction,
      formUrl           : this.props.formUrl,
      closeOnSuccess    : false,
      detailsFetched    : false,
      'task[channel_id]': Object.values(this.props.channels)[0],
      'task[quantity]'  : this.props.task.quantity
    }
  }

  goBack() {
    if (this.state.detailsFetched) {
      this.setState({
        detailsFetched: false,
      })
    } else {
      typeof window === 'undefined' ? null : window.location = document.referrer
    }
  }

  errorAdd(n, e) {
    this.setState({
      errors: Object.assign({}, this.state.errors, {[n]: e})
    })
  }

  errorRemove(n) {
    let e = this.state.errors
    delete e[n]
    this.setState({
      errors: e
    })
  }

  disable(a) {
    let d = this.state.disabled
    a.forEach(n => d = Object.assign({}, d, {[n]: true}))
    this.setState({
      disabled: d
    })
  }

  enable(a) {
    let d = this.state.disabled
    a.forEach(n => delete d[n])
    this.setState({
      disabled: d
    })
  }

  fetchRecipientAddress(channelId, uid, email) {
    this.setState({
      'task[recipient_address]': ''
    })

    fetchPolyfill(this.props.recipientAddressUrl, {
      credentials: 'same-origin',
      method     : 'POST',
      body       : JSON.stringify({
        'channel_id'        : channelId,
        'uid'               : uid,
        'email'             : email,
        'authenticity_token': this.props.csrfToken
      }),
      headers: {
        'Accept'      : 'application/json',
        'Content-Type': 'application/json'
      }
    }).then(response => {
      if (response.status === 200) {
        return response.json()
      } else {
        let errorText = response.text()
        this.setState(state => ({
          flashMessages: state.flashMessages.concat([{'severity': 'error', 'text': errorText}])
        }))
        throw Error(errorText)
      }
    }).then(data => {
      this.setState({
        'task[recipient_address]'    : data.address,
        'task[recipient_address_url]': data.walletUrl,
      })
    })
  }

  handleFieldChange(event) {
    this.setState({ [event.target.name]: event.target.value })

    if (!event.target.checkValidity()) {
      this.errorAdd(event.target.name, 'invalid value')
      return
    } else {
      this.errorRemove(event.target.name)
    }

    if (event.target.value === '') {
      return
    }
  }

  handleSubmit(event) {
    event.preventDefault()

    this.disable(['task[submit]'])

    if (!event.target.checkValidity()) {
      this.enable(['task[submit]'])
      return
    }

    if (!this.state.detailsFetched) {
      this.fetchRecipientAddress(
        this.state['task[channel_id]'],
        this.state['task[uid]'] || this.state['task[channel_id]'] !== '' ? Object.values(this.props.members[this.state['task[channel_id]']])[0] : null,
        this.state['task[email]']
      )

      this.setState({
        detailsFetched: true
      })

      this.enable(['task[submit]'])
      return
    }

    const formData = new FormData(event.target)
    if (this.state['task[channel_id]'] !== '') {
      formData.append('task[channel_id]', this.state['task[channel_id]'] || Object.values(this.props.channels[0]))
      formData.append('task[uid]', this.state['task[uid]'] || Object.values(this.props.members[this.state['task[channel_id]']])[0])
    }

    fetchPolyfill(this.state.formUrl, {
      credentials: 'same-origin',
      method     : this.state.formAction,
      body       : formData,
      headers    : {
        'X-Key-Inflection': 'snake'
      }
    }).then(response => {
      if (response.status === 200) {
        window.location = this.props.urlOnSuccess
      } else {
        response.json().then(data => {
          this.setState(state => ({
            detailsFetched: false,
            errors        : data.errors,
            flashMessages : state.flashMessages.concat([{'severity': 'error', 'text': data.message}])
          }))
          this.enable(['task[submit]'])
        })
      }
    })
  }

  render() {
    return (
      <React.Fragment>
        <Layout
          className="task-award-form"
          title={`Issue Award For Task: ${this.props.task.name}`}
          hasBackButton
          subfooter={
            <React.Fragment>
              <Button
                className="metamask-transfer-btn"
                value={this.state.detailsFetched ? 'issue award' : 'proceed'}
                type="submit"
                form="task-award-form--form"
                disabled={this.state.disabled['task[submit]']}
              />
              <ButtonBorder
                value={this.state.detailsFetched ? 'edit' : 'cancel'}
                onClick={this.goBack}
              />
            </React.Fragment>
          }
        >
          <Flash messages={this.state.flashMessages} />

          <form className="task-award-form--form" id="task-award-form--form" onSubmit={this.handleSubmit}>
            <InputFieldDropdown
              title="communication channel"
              name="task[channel_id]"
              value={this.state['task[channel_id]']}
              errorText={this.state.errors['task[channel_id]']}
              disabled={this.state.detailsFetched || this.state.disabled['task[channel_id]']}
              eventHandler={this.handleFieldChange}
              selectEntries={Object.entries(this.props.channels)}
              symbolLimit={0}
            />

            {this.state['task[channel_id]'] !== '' &&
              <InputFieldDropdown
                title="username"
                name="task[uid]"
                value={this.state['task[uid]'] || Object.values(this.props.members[this.state['task[channel_id]']])[0]}
                errorText={this.state.errors['task[uid]']}
                disabled={this.state.detailsFetched || this.state.disabled['task[uid]']}
                eventHandler={this.handleFieldChange}
                selectEntries={Object.entries(this.props.members[this.state['task[channel_id]']])}
                symbolLimit={0}
              />
            }

            {this.state['task[channel_id]'] === '' &&
              <InputFieldWhiteDark
                title="email"
                type="email"
                required
                name="task[email]"
                value={this.state['task[email]']}
                errorText={this.state.errors['task[email]']}
                readOnly={this.state.detailsFetched || this.state.disabled['task[email]']}
                eventHandler={this.handleFieldChange}
                placeholder="Enter the recepient email address"
                symbolLimit={0}
              />
            }

            <InputFieldDescriptionMiddle
              title="message"
              recommended
              name="task[message]"
              value={this.state['task[message]']}
              errorText={this.state.errors['task[message]']}
              eventHandler={this.handleFieldChange}
              readOnly={this.state.detailsFetched || this.state.disabled['task[message]']}
              placeholder="Give a shoutout or offer some tips to the person who completed the task"
              symbolLimit={150}
            />

            <InputFieldWhiteDark
              title="quantity"
              required
              name="task[quantity]"
              value={this.state['task[quantity]']}
              errorText={this.state.errors['task[quantity]']}
              eventHandler={this.handleFieldChange}
              type="number"
              min="0"
              step="0.01"
              readOnly={this.state.detailsFetched || this.state.disabled['task[quantity]']}
              placeholder="The # of awards to send"
              symbolLimit={0}
            />

            <div className="task-award-form--form--field">
              <div className="task-award-form--form--field--title">
                {`total award amount (${this.props.token.symbol ? this.props.token.symbol : 'TOKEN NOT SPECIFIED'})`}
              </div>
              <div className="task-award-form--form--field--value">
                {
                  Decimal(
                    this.props.task.amount || 0
                  ).toFixed(
                    this.props.token.decimalPlaces || 0,
                    Decimal.ROUND_DOWN
                  )
                }
                {' Ã— '}
                {parseFloat(this.state['task[quantity]'])}
                {' = '}
                {
                  Decimal.mul(
                    this.props.task.amount || 0,
                    parseFloat(this.state['task[quantity]'])
                  ).toDecimalPlaces(
                    this.props.token.decimalPlaces || 0,
                    Decimal.ROUND_DOWN
                  ).toFixed(
                    this.props.token.decimalPlaces || 0,
                    Decimal.ROUND_DOWN
                  )
                }
              </div>
            </div>

            {this.state.detailsFetched &&
              <div className="task-award-form--form--field">
                <div className="task-award-form--form--field--title">
                  recipient address
                </div>
                <div className="task-award-form--form--field--value">
                  {!this.state['task[recipient_address]'] &&
                    <span>
                      The recipient must register their address before they can accept the award.
                    </span>
                  }
                  {this.state['task[recipient_address]'] &&
                    <a href={this.state['task[recipient_address_url]']} target="_blank">
                      {this.state['task[recipient_address]']}
                    </a>
                  }
                </div>
              </div>
            }

            <input
              type="hidden"
              name="authenticity_token"
              value={this.props.csrfToken}
              readOnly
            />
          </form>
        </Layout>
      </React.Fragment>
    )
  }
}
TaskShow.propTypes = {
  task               : PropTypes.object.isRequired,
  batch              : PropTypes.object.isRequired,
  token              : PropTypes.object.isRequired,
  channels           : PropTypes.object.isRequired,
  members            : PropTypes.object.isRequired,
  recipientAddressUrl: PropTypes.string.isRequired,
  formUrl            : PropTypes.string.isRequired,
  formAction         : PropTypes.string.isRequired,
  urlOnSuccess       : PropTypes.string.isRequired,
  csrfToken          : PropTypes.string.isRequired
}
TaskShow.defaultProps = {
  task: {
    'id'         : 28,
    'awardTypeId': 10,
    'name'       : 'Task name',
    'amount'     : '100.0',
    'quantity'   : 1
  },
  batch: {'default': '_'},
  token: {
    'symbol'       : 'DMT',
    'decimalPlaces': 8,
  },
  channels           : {'email': '', 'default': '1'},
  members            : {'1': {'default': '_'}},
  recipientAddressUrl: '/',
  formUrl            : '/',
  formAction         : 'POST',
  urlOnSuccess       : '/',
  csrfToken          : '00'
}
export default TaskShow
