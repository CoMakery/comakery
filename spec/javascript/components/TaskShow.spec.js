import React from 'react'
import { mount } from 'enzyme'
import TaskShow from 'components/TaskShow'

describe('TaskShow', () => {
  it('renders correctly without props', () => {
    const wrapper = mount(<TaskShow />)

    expect(wrapper).toMatchSnapshot()

    expect(wrapper.exists('Layout[title]')).toBe(true)
    expect(wrapper.exists('.task-award-form')).toBe(true)
    expect(wrapper.exists('ButtonBorder[value="cancel"]')).toBe(true)
    expect(wrapper.exists('Button[value="proceed"]')).toBe(true)
    expect(wrapper.exists('.task-award-form--message')).toBe(false)
    expect(wrapper.exists('.task-award-form--form')).toBe(true)
    expect(wrapper.exists('#task-award-form--form')).toBe(true)

    expect(wrapper.exists(
      'InputFieldDropdown[title="communication channel"][name="task[channel_id]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldWhiteDark[title="email"][required][name="task[email]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldDescriptionMiddle[title="message"][recommended][name="task[message]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldWhiteDark[title="quantity"][required][name="task[quantity]"]'
    )).toBe(true)

    expect(wrapper.exists(
      '.task-award-form--form--field'
    )).toBe(true)

    expect(wrapper.exists(
      'input[type="hidden"][name="authenticity_token"]'
    )).toBe(true)
  })

  it('renders correctly with channels and members', () => {
    const channels = {
      'general': '1',
      'random' : '2',
      'email'  : ''
    }
    const members = {
      '1': {
        'slackbot - @slackbot': 'USLACKBOT'
      },
      '2': {
        'slackbot - @slackbot': 'USLACKBOT'
      }
    }

    const wrapper = mount(<TaskShow channels={channels} members={members} />)

    expect(wrapper.find(
      'InputFieldDropdown[name="task[channel_id]"]'
    ).props().value).toBe('1')

    expect(wrapper.find(
      'InputFieldDropdown[name="task[channel_id]"]'
    ).props().selectEntries).toEqual(Object.entries(channels))

    expect(wrapper.find(
      'InputFieldDropdown[name="task[uid]"]'
    ).props().selectEntries).toEqual(Object.entries(members['1']))

    expect(wrapper.exists(
      'InputFieldWhiteDark[title="email"][required][name="task[email]"]'
    )).toBe(false)

    wrapper.find(
      'InputFieldDropdown[name="task[channel_id]"]'
    ).simulate('change', {target: {value: '2'}})

    expect(wrapper.find(
      'InputFieldDropdown[name="task[uid]"]'
    ).props().selectEntries).toEqual(Object.entries(members['2']))

    expect(wrapper.exists(
      'InputFieldWhiteDark[title="email"][required][name="task[email]"]'
    )).toBe(false)
  })

  it('calculates total amount', () => {
    const wrapper = mount(<TaskShow />)

    expect(wrapper.find(
      '.task-award-form--form--field'
    ).text()).toBe('total award amount (DMT)100.00000000 Ã— 1 = 100.00000000')
  })

  it('renders correctly with csrfToken', () => {
    const wrapper = mount(<TaskShow csrfToken="test" />)

    expect(wrapper.find(
      'input[type="hidden"][name="authenticity_token"]'
    ).props().value).toBe('test')
  })

  it('uses formUrl', () => {
    const wrapper = mount(<TaskShow formUrl="/test" />)

    expect(wrapper.state('formUrl')).toBe('/test')
  })

  it('uses formAction', () => {
    const wrapper = mount(<TaskShow formAction="PUT" />)

    expect(wrapper.state('formAction')).toBe('PUT')
  })

  it('uses urlOnSuccess', () => {
    const wrapper = mount(<TaskShow urlOnSuccess="/test" />)

    expect(wrapper.props().urlOnSuccess).toBe('/test')
  })

  it('displays messages', () => {
    const wrapper = mount(<TaskShow />)

    wrapper.setState({
      flashMessages: [
        {
          'severity': 'notice',
          'text'    : 'notice text'
        },
        {
          'severity': 'warning',
          'text'    : 'warning text'
        },
        {
          'severity': 'error',
          'text'    : 'error text'
        }
      ]
    })

    wrapper.update()

    expect(wrapper.exists('Flash')).toBe(true)
  })

  it('displays errors', () => {
    const wrapper = mount(<TaskShow />)

    wrapper.setState({
      errors: {
        'task[channel_id]': 'channel_id error',
        'task[message]'   : 'message error'
      }
    })

    wrapper.update()

    expect(wrapper.exists(
      'InputFieldDropdown[errorText="channel_id error"][name="task[channel_id]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldDescriptionMiddle[errorText="message error"][name="task[message]"]'
    )).toBe(true)
  })
})
