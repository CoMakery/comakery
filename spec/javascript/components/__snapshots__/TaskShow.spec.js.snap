// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskShow renders correctly without props 1`] = `
<TaskShow
  batch={
    Object {
      "default": "_",
    }
  }
  channels={
    Object {
      "default": "1",
      "email": "",
    }
  }
  csrfToken="00"
  formAction="POST"
  formUrl="/"
  members={
    Object {
      "1": Object {
        "default": "_",
      },
    }
  }
  project={
    Object {
      "default": "_",
    }
  }
  recipientAddressUrl="/"
  task={
    Object {
      "amount": "100.0",
      "awardTypeId": 10,
      "id": 28,
      "name": "Task name",
      "quantity": 1,
    }
  }
  token={
    Object {
      "decimalPlaces": 8,
      "symbol": "DMT",
    }
  }
  urlOnSuccess="/"
>
  <ProjectSetup
    className="task-award-form"
    editable={true}
    hasBackButton={true}
    projectId={null}
    projectPage="batches"
    projectTitle=""
    sidebar={
      <React.Fragment>
        <div
          className="batch-index--sidebar"
        >
          <SidebarItem
            className="batch-index--sidebar--item batch-index--sidebar--item__form"
            iconLeftName=""
            iconLeftUrl=""
            iconRightName=""
            notificationColor=""
            selected={true}
            text="sidebar item"
          />
          <hr
            className="batch-index--sidebar--hr"
          />
        </div>
      </React.Fragment>
    }
    subfooter={
      <React.Fragment>
        <Button
          className="metamask-transfer-btn"
          disabled={false}
          form="task-award-form--form"
          type="submit"
          value="proceed"
        />
        <ButtonBorder
          className=""
          onClick={[Function]}
          value="cancel"
        />
      </React.Fragment>
    }
  >
    <Layout
      category=""
      className="project-setup task-award-form"
      customTitle={
        <ProjectSetupHeader
          projectId={null}
          projectOwner={true}
          projectPage="batches"
          projectTitle=""
        />
      }
      hasBackButton={true}
      sidebar={
        <React.Fragment>
          <div
            className="batch-index--sidebar"
          >
            <SidebarItem
              className="batch-index--sidebar--item batch-index--sidebar--item__form"
              iconLeftName=""
              iconLeftUrl=""
              iconRightName=""
              notificationColor=""
              selected={true}
              text="sidebar item"
            />
            <hr
              className="batch-index--sidebar--hr"
            />
          </div>
        </React.Fragment>
      }
      subfooter={
        <React.Fragment>
          <Button
            className="metamask-transfer-btn"
            disabled={false}
            form="task-award-form--form"
            type="submit"
            value="proceed"
          />
          <ButtonBorder
            className=""
            onClick={[Function]}
            value="cancel"
          />
        </React.Fragment>
      }
      title={null}
    >
      <div
        className="layout project-setup task-award-form"
      >
        <div
          className="layout--back-button"
        >
          <Icon
            className=""
            name="iconCloseCopy.svg"
            onClick={[Function]}
          >
            <img
              className="icon icon__iconCloseCopy-svg"
              onClick={[Function]}
              src="test-file-stub"
            />
          </Icon>
        </div>
        <div
          className="layout--content"
        >
          <div
            className="layout--content--title"
          >
            <ProjectSetupHeader
              projectId={null}
              projectOwner={true}
              projectPage="batches"
              projectTitle=""
            >
              <div
                className="layout--content--title--header"
              />
              <span
                className="layout--content--title--nav"
              >
                <a
                  href="#"
                >
                  project settings
                </a>
              </span>
              <span
                className="layout--content--title--nav__current"
              >
                <a
                  href="#"
                >
                  batches & tasks
                </a>
              </span>
              <span
                className="layout--content--title--nav"
              >
                <a
                  href="#"
                >
                  contributors
                </a>
              </span>
              <span
                className="layout--content--title--nav"
              >
                <a
                  href="#"
                >
                  payments
                </a>
              </span>
              <span
                className="layout--content--title--nav"
              >
                <a
                  href="#"
                >
                  overview
                </a>
              </span>
            </ProjectSetupHeader>
          </div>
          <hr
            className="layout--content--hr"
          />
          <div
            className="layout--content--wrapper"
          >
            <div
              className="layout--content--sidebar"
            >
              <div
                className="batch-index--sidebar"
              >
                <SidebarItem
                  className="batch-index--sidebar--item batch-index--sidebar--item__form"
                  iconLeftName=""
                  iconLeftUrl=""
                  iconRightName=""
                  notificationColor=""
                  selected={true}
                  text="sidebar item"
                >
                  <div
                    className="sidebar-item--wrapper"
                  >
                    <div
                      className="sidebar-item sidebar-item__selected batch-index--sidebar--item batch-index--sidebar--item__form"
                    >
                      <div
                        className="sidebar-item--content"
                      >
                        <div
                          className="sidebar-item--text"
                        >
                          sidebar item
                        </div>
                        <div
                          className="sidebar-item--subchild"
                        />
                      </div>
                      <div
                        className="sidebar-item--left-border"
                      />
                      <div
                        className="sidebar-item--right-border sidebar-item--right-border__"
                      />
                    </div>
                  </div>
                </SidebarItem>
                <hr
                  className="batch-index--sidebar--hr"
                />
              </div>
            </div>
            <div
              className="layout--content--content__sidebared"
            >
              <Flash
                className=""
                messages={Array []}
              >
                <div
                  className="flash-message-container"
                />
              </Flash>
              <div
                className="task-form--form--title"
              >
                Issue Award For Task: Task name
              </div>
              <form
                className="task-award-form--form"
                id="task-award-form--form"
                onSubmit={[Function]}
              >
                <InputFieldDropdown
                  eventHandler={[Function]}
                  name="task[channel_id]"
                  selectEntries={
                    Array [
                      Array [
                        "email",
                        "",
                      ],
                      Array [
                        "default",
                        "1",
                      ],
                    ]
                  }
                  symbolLimit={0}
                  title="communication channel"
                  value=""
                >
                  <InputField
                    checkboxText=""
                    checked={false}
                    className="input-field__dropdown"
                    copyOnClick={false}
                    disabled={false}
                    errorText=""
                    eventHandler={[Function]}
                    imgInputRef={null}
                    imgPreviewDimensions="40x40"
                    imgPreviewUrl=""
                    imgRequirements=""
                    imgVerifier={[Function]}
                    max=""
                    min=""
                    name="task[channel_id]"
                    pattern=".*"
                    placeholder="Please enter value"
                    readOnly={false}
                    recommended={false}
                    required={false}
                    selectEntries={
                      Array [
                        Array [
                          "email",
                          "",
                        ],
                        Array [
                          "default",
                          "1",
                        ],
                      ]
                    }
                    step=""
                    symbolLimit={0}
                    title="communication channel"
                    type="select"
                    value=""
                  >
                    <div
                      className="input-field input-field__dropdown"
                    >
                      <div
                        className="input-field--title"
                      >
                        <span
                          className="input-field--title--title"
                        >
                          communication channel
                        </span>
                        <span
                          className="input-field--title--required"
                        >
                          optional
                        </span>
                      </div>
                      <div
                        className="input-field--content"
                      >
                        <select
                          className="input-field--content__select"
                          disabled={false}
                          name="task[channel_id]"
                          onChange={[Function]}
                          required={false}
                          value=""
                        >
                          <option
                            key=""
                            value=""
                          >
                            email
                          </option>
                          <option
                            key="1"
                            value="1"
                          >
                            default
                          </option>
                        </select>
                      </div>
                    </div>
                  </InputField>
                </InputFieldDropdown>
                <InputFieldWhiteDark
                  eventHandler={[Function]}
                  name="task[email]"
                  placeholder="Enter the recepient email address"
                  required={true}
                  symbolLimit={0}
                  title="email"
                  type="email"
                >
                  <InputField
                    checkboxText=""
                    checked={false}
                    className="input-field__white__dark"
                    copyOnClick={false}
                    disabled={false}
                    errorText=""
                    eventHandler={[Function]}
                    imgInputRef={null}
                    imgPreviewDimensions="40x40"
                    imgPreviewUrl=""
                    imgRequirements=""
                    imgVerifier={[Function]}
                    max=""
                    min=""
                    name="task[email]"
                    pattern=".*"
                    placeholder="Enter the recepient email address"
                    readOnly={false}
                    recommended={false}
                    required={true}
                    selectEntries={Array []}
                    step=""
                    symbolLimit={0}
                    title="email"
                    type="email"
                    value=""
                  >
                    <div
                      className="input-field input-field__white__dark"
                    >
                      <div
                        className="input-field--title"
                      >
                        <span
                          className="input-field--title--title"
                        >
                          email
                        </span>
                        <span
                          className="input-field--title--required"
                        >
                          required
                        </span>
                      </div>
                      <div
                        className="input-field--content"
                      >
                        <input
                          className="input-field--content__text"
                          name="task[email]"
                          onChange={[Function]}
                          onClick={null}
                          pattern=".*"
                          placeholder="Enter the recepient email address"
                          readOnly={false}
                          required={true}
                          type="email"
                          value=""
                        />
                      </div>
                    </div>
                  </InputField>
                </InputFieldWhiteDark>
                <InputFieldDescriptionMiddle
                  eventHandler={[Function]}
                  name="task[message]"
                  placeholder="Give a shoutout or offer some tips to the person who completed the task"
                  recommended={true}
                  symbolLimit={150}
                  title="message"
                >
                  <InputField
                    checkboxText=""
                    checked={false}
                    className="input-field__description-middle"
                    copyOnClick={false}
                    disabled={false}
                    errorText=""
                    eventHandler={[Function]}
                    imgInputRef={null}
                    imgPreviewDimensions="40x40"
                    imgPreviewUrl=""
                    imgRequirements=""
                    imgVerifier={[Function]}
                    max=""
                    min=""
                    name="task[message]"
                    pattern=".*"
                    placeholder="Give a shoutout or offer some tips to the person who completed the task"
                    readOnly={false}
                    recommended={true}
                    required={false}
                    selectEntries={Array []}
                    step=""
                    symbolLimit={150}
                    title="message"
                    type="textarea"
                    value=""
                  >
                    <div
                      className="input-field input-field__description-middle"
                    >
                      <div
                        className="input-field--title"
                      >
                        <span
                          className="input-field--title--title"
                        >
                          message
                        </span>
                        <span
                          className="input-field--title--required"
                        >
                          recommended
                        </span>
                        <span
                          className="input-field--title--counter"
                        >
                          0
                          /
                          150
                        </span>
                      </div>
                      <div
                        className="input-field--content"
                      >
                        <textarea
                          className="input-field--content__text"
                          name="task[message]"
                          onChange={[Function]}
                          pattern=".*"
                          placeholder="Give a shoutout or offer some tips to the person who completed the task"
                          readOnly={false}
                          required={false}
                          value=""
                        />
                      </div>
                    </div>
                  </InputField>
                </InputFieldDescriptionMiddle>
                <InputFieldWhiteDark
                  eventHandler={[Function]}
                  min="0"
                  name="task[quantity]"
                  placeholder="The # of awards to send"
                  required={true}
                  step="0.01"
                  symbolLimit={0}
                  title="quantity"
                  type="number"
                  value={1}
                >
                  <InputField
                    checkboxText=""
                    checked={false}
                    className="input-field__white__dark"
                    copyOnClick={false}
                    disabled={false}
                    errorText=""
                    eventHandler={[Function]}
                    imgInputRef={null}
                    imgPreviewDimensions="40x40"
                    imgPreviewUrl=""
                    imgRequirements=""
                    imgVerifier={[Function]}
                    max=""
                    min="0"
                    name="task[quantity]"
                    pattern=".*"
                    placeholder="The # of awards to send"
                    readOnly={false}
                    recommended={false}
                    required={true}
                    selectEntries={Array []}
                    step="0.01"
                    symbolLimit={0}
                    title="quantity"
                    type="number"
                    value={1}
                  >
                    <div
                      className="input-field input-field__white__dark"
                    >
                      <div
                        className="input-field--title"
                      >
                        <span
                          className="input-field--title--title"
                        >
                          quantity
                        </span>
                        <span
                          className="input-field--title--required"
                        >
                          required
                        </span>
                      </div>
                      <div
                        className="input-field--content"
                      >
                        <input
                          className="input-field--content__text"
                          max=""
                          min="0"
                          name="task[quantity]"
                          onChange={[Function]}
                          onClick={null}
                          placeholder="The # of awards to send"
                          readOnly={false}
                          required={true}
                          step="0.01"
                          type="number"
                          value={1}
                        />
                      </div>
                    </div>
                  </InputField>
                </InputFieldWhiteDark>
                <div
                  className="task-award-form--form--field"
                >
                  <div
                    className="task-award-form--form--field--title"
                  >
                    total award amount (DMT)
                  </div>
                  <div
                    className="task-award-form--form--field--value"
                  >
                    100.00000000
                     × 
                    1
                     = 
                    100.00000000
                  </div>
                </div>
                <input
                  name="authenticity_token"
                  readOnly={true}
                  type="hidden"
                  value="00"
                />
              </form>
            </div>
          </div>
        </div>
        <div
          className="layout--subfooter"
        >
          <div
            className="layout--subfooter--buttons"
          >
            <Button
              className="metamask-transfer-btn"
              disabled={false}
              form="task-award-form--form"
              type="submit"
              value="proceed"
            >
              <input
                className="button metamask-transfer-btn"
                disabled={false}
                form="task-award-form--form"
                type="submit"
                value="proceed"
              />
            </Button>
            <ButtonBorder
              className=""
              onClick={[Function]}
              value="cancel"
            >
              <Button
                className="button__border"
                disabled={false}
                onClick={[Function]}
                type="button"
                value="cancel"
              >
                <input
                  className="button button__border"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                  value="cancel"
                />
              </Button>
            </ButtonBorder>
          </div>
        </div>
      </div>
    </Layout>
  </ProjectSetup>
</TaskShow>
`;
